redis/client.py

```diff
@@ -232,6 +232,61 @@ def int_or_none(response):
     return int(response)
 
 
+def stream_key(response):
+    return response
+
+
+def stream_list(response):
+    if response is None:
+        return None
+    result = []
+    for r in response:
+        kv_pairs = r[1]
+        kv_dict = dict()
+        while len(kv_pairs) > 1:
+            kv_dict[kv_pairs.pop()] = kv_pairs.pop()
+        result.append((r[0], kv_dict))
+
+    return result
+
+
+def parse_recursive_dict(response):
+    if response is None:
+        return None
+    result = {}
+    while response:
+        k = response.pop(0)
+        v = response.pop(0)
+        if isinstance(v, list):
+            v = parse_recursive_dict(v)
+        result[k] = v
+    return result
+
+
+def parse_list_of_recursive_dicts(response):
+    if response is None:
+        return None
+    result = []
+    for group in response:
+        result.append(parse_recursive_dict(group))
+    return result
+
+
+def parse_xclaim(response):
+    if isinstance(response, str):
+        return response
+    return stream_list(response)
+
+
+def multi_stream_list(response):
```

Comment:
There are several options for the return structure of `multi_stream_list`:

```python
# a list of of 2-item stream/messages lists
[
    [
        'stream-1',
        [
            [
                '1540456647578-0',
                {'key-1': 'value-1', 'key-2': 'value-2'}
            ],
            [
                '1540456647578-1',
                {'key-1': 'value-1', 'key-2': 'value-2'}
            ],
        ],
    ],
    [
        'stream-2',
        [
            [
                '1540456647578-0',
                {'key-1': 'value-1', 'key-2': 'value-2'}
            ],
            [
                '1540456647578-1',
                {'key-1': 'value-1', 'key-2': 'value-2'}
            ],
        ],
    ],
]

# this allows all list-based iterating....
for stream, messages in r.xread(streams={'stream-1': 0, 'stream-2': 0}):
    for message_id, message_data in messages:
        ....


# alternatively, the existing code represents this return value
# as a dict of streams to lists of message id: dict attr pairs
{
    'stream-1': [
        [
            '1540456647578-0',
            {'key-1': 'value-1', 'key-2': 'value-2'}
        ],
        [
            '1540456647578-1',
            {'key-1': 'value-1', 'key-2': 'value-2'}
        ],
    ],
    'stream-2': [
        [
            '1540456662589-0',
            {'key-1': 'value-1', 'key-2': 'value-2'}
        ],
        [
            '1540456662589-1',
            {'key-1': 'value-1', 'key-2': 'value-2'}
        ],
    ]
}

# iterating here requires the extra .items() call
for stream, messages in r.xread(streams={'stream-1': 0, 'stream-2': 0}).items()
    for message_id, message_data in messages:
        ....
# given redis returns null when there are no messages to read,
# you can't reliably index directly into the dictionary, which seems to
# take away much of the value of the dictionary
# this blows up with a TypeError if there's no message to read
# since xread()'s return value will be None.
r.xread(streams={'stream-1': 99999999999})['stream-1']
```

I think the first option might be the best choice here. I don't see any value from making the return value a dict of {stream: messages} pairs when you can't reliably index into the dictionary. It seems like this just forces users to type more. Am I missing something?

Reply:
ok. 
allow list based iterating on XREADGROUP results